metal/ringbuf.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#ifndef METAL__RINGBUF_H
+#define METAL__RINGBUF_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*!
+ * @brief The handle for a ring buffer
+ */
+struct metal_ringbuf {
+    /* _buf points at the base of the ring buffer memory */
+    void *_buf;
+    /* _start points at the ring buffer location containing the first value to
+     * be read by metal_ringbuf_get */
+    void *_start;
+    /* _end points at the first empty location after the buffer contents */
+    void *_end;
+    /* _capacity is the number of items the ring buffer can hold */
+    size_t _capacity;
+    /* _item_size is the size of each element in the ring buffer memory */
+    size_t _item_size;
+    /* _empty is 0 when the buffer is empty, 1 otherwise. This is used to
+     * differentiate
+     * the full condition from the empty condition, for which both cases have
+     * _start == _end */
+    bool _empty;
```

Comment:
You can avoid the separate _empty boolean by using a 'count' value that tracks how much data is in the ring instead of  '_start'.
If you do choose to do this, I'd suggest using an integer for _end too. Oh, names are hard; _end and _count are not exactly clear. I've used 'read' and 'count' to indicate the place to start reading and the count of data in the buffer, but you might have better ideas.